### ISSUE TITLE
[feature request] Conv3D vol2col and col2vol optimization with CPU backend

### ISSUE THREAD

**Issue by MlWoo** [pytorch/pytorch#4738]
It is strange that vol2col and col2vol in Conv3D could not be parallelized with [CPU backend](https://github.com/pytorch/pytorch/blob/master/aten/src/THNN/generic/VolumetricConvolutionMM.c#L114). Actually `THNN_(unfolded_copy_vol)` function is OK to use OpenMP. But test case will fail when using OpenMP in `THNN_(unfolded_acc_vol)` function . The reason why the transforms could not be parallized with CPU backend is  read-write confict with multi-thread.  However, it will avert the problem if we adopt another coding strategy as that in [GPU backend](https://github.com/pytorch/pytorch/blob/master/aten/src/THCUNN/VolumetricConvolution.cu#L9).  Is it OK to optimize the code?

**Comment by apaszke** [comment]
Just out of curiosity how does #2764 allow us to use SIMD for `exp` or `log`? As far as I understood, it only improves some thread-level paralellism, not vectorization. For example the inner loop of `sum` in that PR still can't be vectorized because we compile without `-ffast-math`, so the fact that we're using a single accumulator prevents this optimization. Remember that the results you're getting with `icc` (which has this flag on by default IIRC) might not be representable of the assembly that `gcc` and `clang` produce, but these are the compilers that 99.99% of our users use (since they're free). In fact, we're only using `gcc` to build our binaries.

**Comment by fmassa** [comment]
@apaszke I think SIMD vectorization in #2764 might come from https://github.com/pytorch/pytorch/pull/2764/files#diff-0898bb4475bffe9488f7e8dc5fd80ff4R375

But I'm still suspicious why my basic benchmarks with numpy were giving such great performances even though they were single threaded.

@MlWoo about your question on conv3d, I think this relates a bit with https://github.com/torch/nn/issues/501. We would need to recompute the `finput` buffer for the gradient accumulation, and that might take some extra time. On the other hand, having a huge buffer for conv3d dependent on the batch size is not great because it is much more memory hungry than the one in conv2d.

**Comment by apaszke** [comment]
That makes sense, but I'm pretty sure neither `gcc` nor `clang` have vectorized intrinsics for `log` or `exp`, so it's an `icc` only optimization.

**Comment by apaszke** [comment]
Also, the fact that we define a macro, which now suddenly asserts that there are no loop carried dependencies makes me a bit worried 😕

**Comment by MlWoo** [comment]
@apazke, maybe all improvements comes from openmp. But we have not enabled icc compilation on pytorch because too many cmake files should be modified. The version which do 6.6x boost is compiled by gcc contained in anaconda.

**Comment by apaszke** [comment]
Ok, that's great news! But I don't think OpenMP does anything to vectorize `log` or `exp`. I have never seen it do that to other parts of our code, so some extra speedups might be very easy to achieve.

**Comment by MlWoo** [comment]
@apaszke I think [ivdep](https://github.com/intel/pytorch/blob/dev-omp/aten/src/TH/THTensorApply.h#L375) and [simd] pragma directive will hint the compiler to vectorize many math operations like `exp` and `log` when the operands are contiguous.

**Comment by mingfeima** [comment]
MlWoo brought up the conv3d issue because we received formal request from HPC researchers who are using super computers (mostly CPU only) in [high energy physics domain](http://www.nersc.gov/users/data-analytics/data-analytics-2/deep-learning/). These researchers are currently using TF but they have a strong wish to shift to pytorch as they claimed dynamic graph is more easy to construct their model. We helped them deploy pytorch on super computer environment and did some optimization. Conv3d is one of their hot features, so Mlwoo did the optimization.

`log` or `exp` can be properly vectorized by using svml with gcc, -mveclibabi=svml, icc does the vectorization by default.

**Comment by apaszke** [comment]
@mingfeima We're ok with any improvements you send, I'm just pointing out there might be simpler and more impactful changes. Is SVML publicly available for free?

**Comment by MlWoo** [comment]
@apaszke Could you help me review the PR #4824 ?
